cmd/ido2icingadb/misc.go

```diff
@@ -0,0 +1,336 @@
+		if err != nil {
+			log.With("query", query).Fatalf("%+v", errors.Wrap(err, "can't prepare query"))
+		}
+
+		var rows []Row
+		if err := stmt.Select(&rows, args); err != nil {
+			log.With("query", query).Fatalf("%+v", errors.Wrap(err, "can't perform query"))
+		}
+
+		_ = stmt.Close()
+
+		if len(rows) < 1 {
+			break
+		}
+
+		if checkpoint = onRows(rows); checkpoint == nil {
+			break
+		}
+
+		args["checkpoint"] = checkpoint
+	}
+}
+
+// historyType specifies a history data type.
+type historyType struct {
+	// name is a human-readable, but machine-friendly common name.
+	name string
+	// idoTable specifies the source table.
+	idoTable string
+	// idoIdColumn specifies idoTable's primary key.
+	idoIdColumn string
+	// cacheSchema specifies <name>.sqlite3's structure.
+	cacheSchema []string
+	// cacheFiller fills cache from snapshot.
+	cacheFiller func(*historyType)
+	// cacheLimitQuery rationale: see migrate().
+	cacheLimitQuery string
+	// migrationQuery SELECTs source data for actual migration.
+	migrationQuery string
+	// migrate does the actual migration.
+	migrate func(c *Config, idb *icingadb.DB, envId []byte,
+		endpointId [sha1.Size]byte, idbTx *sync.Mutex, ht *historyType)
+
+	// cache represents <name>.sqlite3.
+	cache *sqlx.DB
+	// snapshot represents the data source.
+	snapshot *sqlx.Tx
+	// total summarizes the source data.
+	total int64
+	// done summarizes the migrated data.
+	done int64
+	// bar represents the current progress bar.
+	bar *mpb.Bar
+	// lastId is the last already migrated ID.
+	lastId uint64
+}
+
+// setupBar (re-)initializes ht.bar.
+func (ht *historyType) setupBar(progress *mpb.Progress) {
+	e := &eta{WC: decor.WC{W: 4}}
+
+	e.Init()
+
+	ht.bar = progress.AddBar(
+		ht.total,
+		mpb.BarFillerClearOnComplete(),
+		mpb.PrependDecorators(
+			decor.Name(ht.name, decor.WC{W: len(ht.name) + 1, C: decor.DidentRight}),
+			decor.Percentage(decor.WC{W: 5}),
+		),
+		mpb.AppendDecorators(e),
+	)
+}
+
+type historyTypes [6]historyType
+
+// forEach performs f per *ht in parallel.
+func (ht *historyTypes) forEach(f func(*historyType)) {
+	eg, _ := errgroup.WithContext(context.Background())
+	for i := range *ht {
+		i := i
+		eg.Go(func() error {
+			f(&(*ht)[i])
+			return nil
+		})
+	}
+
+	_ = eg.Wait()
+}
+
+var types = historyTypes{
+	{
+		name:        "acknowledgement",
+		idoTable:    "icinga_acknowledgements",
+		idoIdColumn: "acknowledgement_id",
+		cacheSchema: eventTimeCacheSchema,
+		cacheFiller: func(ht *historyType) {
+			buildEventTimeCache(ht, []string{
+				"xh.acknowledgement_id id", "UNIX_TIMESTAMP(xh.entry_time) event_time",
+				"xh.entry_time_usec event_time_usec", "xh.acknowledgement_type event_is_start", "xh.object_id",
+			})
+		},
+		migrationQuery: acknowledgementMigrationQuery,
+		migrate: func(c *Config, idb *icingadb.DB, envId []byte, endpId [20]byte, idbTx *sync.Mutex, ht *historyType) {
+			migrateOneType(c, idb, envId, endpId, idbTx, ht, convertAcknowledgementRows)
+		},
```

Comment:
In general, there are so many levels of callbacks in the code that it's hard to understand. For example, this sets a callback as an attributes that calls a function with a callback as an argument which in turn takes a callback as an argument.

For now, I don't have a super concrete suggestion like for `chunkCacheTx`, but it should also be possible to simplify this significantly.